### PR TITLE
[serve] Add JSON schemas for REST API

### LINKED ISSUE #21879: [Feature] [Serve] [Declarative API] Configuration schema
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

Add a Pydantic-validated schema to Serve as an internal representation for all options in the declarative API's configuration YAML file and REST request bodies.

### Use case

The declarative API allows users to specify configuration options for their Serve deployments. This feature adds a Pydantic-validated schema to Serve as an internal representation for these options.

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

This change adds JSON schemas for Serve’s REST API.

## Related issue number

Closes #21879.

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
      - This change adds unit tests that assess all the schemas' behavior.

### CODE REVIEW

File: dashboard/modules/serve/schema.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+        default=None,
+        description=(
+            "Forces replicas to run on nodes with the specified accelerator type."
+        ),
+    )
+    max_concurrency: int = Field(default=None, description=(""))
+    max_task_retries: int = Field(default=None, description=(""))
+    max_pending_calls: int = Field(default=None, description=(""))
+    placement_group_bundle_index: int = Field(default=None, description=(""))
+    placement_group_capture_child_tasks: bool = Field(default=None, description=(""))
```
Comment (edoakes) [inline dashboard/modules/serve/schema.py:244]: isn't related to resources

**Comment by edoakes** [comment]
Gave feedback offline, lmk when it's ready for re-review

**Review by edoakes — APPROVED** [review]

**Comment by edoakes** [comment]
test_output failing on maste

**Comment by dmtrs** [comment]
> @shrekris-anyscale where are the original json schema files (`*.json`) containing the schema definitions located? Are they available online in github?

Nevermind, I was under the impression a json schema was used as input to create the Models. I was able to get the jsons chema using this:
```python
import json
from pydantic import BaseModel
from pydantic.schema import schema

from inspect import getmembers

import ray.serve.schema

title = f"ray.serve.schema@{ray.__version__}"
models = []
for [_, M] in getmembers(ray.serve.schema):
    try:
        if issubclass(M, BaseModel) and M is not BaseModel:
            models.append(M)
    except:
        continue

top_level_schema = schema(models, title=title)
with open(f"./json_schema.{title}.json", 'w') as f:
    f.write(json.dumps(top_level_schema, indent=2))
```

Now I am planning to use json schema to produce typescript sdk for ray servve to be used as clienet. I can see there is no such package on the npm registry https://www.npmjs.com/search?q=ray-project%20ray%20serve 

I could consider contributing as community based package.

**Comment by dmtrs** [comment]
I can also see that there are type definitions in https://github.com/ray-project/ray/blob/master/dashboard/client/src/type/serve.ts but this are not dedicated package along with some http client probably. And also seems file is maintained as features implemented on the dashboard 🤔 and not holistic as what API offers (what I am looking for).

**Pull request merged**